### DISCUSSION TITLE
Building an in-app web3 browser, can WalletConnect help?

### DISCUSSION THREAD

**Discussion by aroth** [WalletConnect/walletconnect-monorepo#967] [Developer Support] [answered]
We're building an Ethereum & Solana wallet app using React Native. One of the features we'd like to integrate is an in-app browser so that Dapps can be used right within the app (ex: browse to OpenSea.io and it detects the your wallet address automatically much like the Eth Wallet chrome extensions work). This is similar to what AlphaWallet, Slope Wallet, Trust, etc... all have as part of their iOS apps. 

Problem is-- where to begin with this? What is required to make this work? What are these other wallets injecting into the in-app browser? For all of our programming expertise, we have very limited context on how extensions and, similarly, in-app web3 browsers are working. Any guidance would be appreciated.

That said, are any of the WalletConnect APIs what we're looking for here?

**Comment by cmrnfaith** [comment]
I'm not a lead developer here, but I know that most web-based wallets talk to the page via a global "ethereum" variable. If you haven't figured this out yet I would take a look into how to call meta mask functions through web3js.

**Comment by hackyguru** [comment, marked as answer]
In your case, WalletConnect Web3Provider might be useful. Refer more at https://docs.walletconnect.com/quick-start/dapps/web3-provider . Hope this helps.

**Comment by KirillRodichevUtorg** [comment]
@aroth I am working on a similar task and would appreciate your expertise on integrating an in-app browser with dApp connection capabilities. Could you share your insights?
